File: 102-binary-tree-level-order-traversal /main.py
import collections
from typing import List, Optional


class TreeNode:
	def __init__(self, val = 0, left = None, right = None):
		self.val = val
		self.left = left
		self.right = right

class Solution:
	def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
		res = []
		q = collections.deque()

		q.append(root)

		while q:
			# this qLen ensures we iterate through one level at a time.
			qLen = len(q)
			level = []

			for i in range(qLen):
				node = q.popleft()

				# the node could be null
				if node:
					level.append(node.val)

					# this and the right node could be null, but that's why we wrap them into an if block
					q.append(node.left)
					q.append(node.right)

			# after we're done with this level, add the current level to the list of levels(res).
			# Note: our queue could have null nodes, we don't want to add them to the res(we initialized the level as an empty
			# list, but we only add it to the res if current level wasn't empty, so we need this if block)
			if level:
				res.append(level)

		return res

File: 102-binary-tree-level-order-traversal /main.js
// Definition for a binary tree node.
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
    const queue = [];
    const res = [];

    if (root) {
        queue.push(root)
    }

    while (queue.length) {
        const queueLength = queue.length // get the length pf queue before adding the left and right nodes to it
        const currLevel = []

        for (let i = 0; i < queueLength; i++) {
            const current = queue.shift()
            currLevel.push(current.val)

            if (current.left) {
                queue.push(current.left)
            }

            if (current.right) {
                queue.push(current.right)
            }
        }

        res.push(currLevel)
    }

    return res
};